ARCHITECTURE.md

```diff
@@ -0,0 +1,456 @@
+# Architecture
+
+This document describes the high-level architecture of PyMC3.
+If you want to familiarize yourself with the code base, you are just in the right place!
+
+
+# Bird's Eye View
+
+TODO: Update diagram to include SMC
+![Architecture](docs/Architecture.png)
+
+On the highest level, PyMC3 lets you define probabilistic graphs or models
+that can be easily used to compute log probabilities for posterior inference
+or to draw random samples from priors or posteriors. 
+However there are numerous
+other useful sets of functionality, from prior and posterior predictive sampling,
+GraphviZ model, as well as various helper functions.
+
+PyMC3 includes a few inference techniques, in particular:
+* Markov chain Monte Carlo 
+* Variational Inference
+* Sequential Monte Carlo
+
+The most relevant modules of PyMC3 are shown visually below, as well as their
+relation to other major packages. Not all modules are shown, either because
+they are smaller or self explanatory in scope, or they're pending
+deprecation
+
+
+## Functionality not in PyMC3
+It is easier to start with functionality that is not present in PyMC3 but
+rather deferred to outside libraries. If seeking to understand any 
+of the topics below refer to that specific library
+
+### Aesara
+* Gradient computation
+* Random number generation
+* Low level tensor operation definition
+* Low level operation graphs
+
+### ArviZ
+* Plotting e.g. Trace plots, rank plots, posterior plots
+* MCMC sampling diagnostics e.g. Rhat, Effective Sample Size.
+* Model comparison, particularly efficient leave-one-out cross-validation approximation
+* Inference Data structure
+
```

Comment:
maybe also add some mentions of honor for bambi/sunode/...? 

They are more user-facing than dev-facing, and I don't think that there is any part of the codebase needing or being specific to them, but it could be helpful in 1) understanding why there isn't a glm module and 2) seeing if some few feature should be a new module or a new library instead.

Reply:
Leaving this open as maybe, I think those are related libraries but dont directly affect the architecture. Aesara for sure does, ArviZ does as well since its a hard requirement of pymc

Reply:
Been thinking about this more. Theres two things
* In regards to related packages  I dont think architecture.md is the right place for this. There shoould be perhaps some other page that links out to all the things built on top of pymc3 like xarray does here http://xarray.pydata.org/en/stable/ecosystem.html
* In regards to "why things arent here" like glm, I dont think this is useful either, especially if we have the above doc

Reply:
Yeah, I'm not sure if here is the best place, but I feel we should have something related to this somewhere. Not so much on "why things aren't there" but on how to decide if a new feature should be a new module or a new library. If there are any other0 than developer convenience

A 3rd party packages/ecosyste. page like matplotlib or xarray have would be amazing cc @martinaca

Reply:
I agree that figuring out whether something should be new library or new module is an architectural decision. I have two questions when thinking about this

1. Given half of the answer is dependent on what the other thing is, would we be able to effectively answer that question in this doc.
2. Should this document just be what the existing architecture is, and not attempt to answer how it should evolve, or what it isnt?

I really dont have answers to the above question but curious to hear anything